storagemarket/deal_handler.go

```diff
@@ -2,31 +2,45 @@ package storagemarket
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
 
 	"github.com/filecoin-project/boost/storagemarket/types"
 	"github.com/libp2p/go-eventbus"
 	"github.com/libp2p/go-libp2p-core/event"
-
-	"github.com/google/uuid"
 )
 
+type transferCancelResp struct {
+	err error // err=nil means that the transfer cancellation was successful.
+}
+
 // dealHandler keeps track of the deal while it's executing
 type dealHandler struct {
+	ctx context.Context
+
+	// Transfer synchronization state
+	transferWriteOnce    sync.Once
+	transferCtx          context.Context
+	transferCancel       context.CancelFunc
+	transferredCancelled chan transferCancelResp // channel the deal execution go-routine will write to when transfer is stopped.
+
 	dealUuid uuid.UUID
-	ctx      context.Context
-	stop     context.CancelFunc
-	stopped  chan struct{}
 	bus      event.Bus
 }
 
-func (d *dealHandler) cancel(ctx context.Context) {
-	d.stop()
+func (d *dealHandler) cancelTransfer(ctx context.Context) error {
+	d.transferCancel()
 	select {
 	case <-ctx.Done():
-		return
-	case <-d.stopped:
-		return
+		return ctx.Err()
+	case tc, ok := <-d.transferredCancelled:
+		if !ok {
+			return errors.New("cannot cancel transfer anymore")
+		}
+		return tc.err
```

Comment:
IIUC there are two ways that an error can be returned here: because the channel has already been closed, or because transferCancelled returns a struct with a non-nil error field, which is also because the transfer can no longer be cancelled. It's quite hard to follow how we end up in these states. Can we simplify?